### PR TITLE
Make cilium pprof listen address configurable

### PR DESCRIPTION
**Pull request by chancez**
Cilium's profiles cannot be gathered by other tools running outside the cilium pod if it's only listening on localhost. To enable continuous profiling with tools like Parca or Phlare which operate on a scrape model, we should make the pprof listen address configurable.

```release-note
Make cilium pprof listen address configurable
```

### CODE REVIEW

**Review by christarazi — APPROVED** [review]
:rocket:

**Comment by christarazi** [comment]
Travis is pointing out that the cmdref needs to be updated. `make -C Documentation/ update-cmdref`, ensure that you have all binaries built first though.

**Comment by chancez** [comment]
@christarazi I'm on MacOS so that's tricky, the scripts all assume you've got everything built locally. Guess I'll have to setup a dev env inside my VM (I normally build everything via docker build).

**Review by joestringer — APPROVED** [review]

File: install/kubernetes/cilium/values.yaml [hunk 1]
```diff
@@ -1531,8 +1531,12 @@ nodePort:
 policyEnforcementMode: "default"
 
 pprof:
-  # -- Enable Go pprof debugging
+  # -- Enable pprof for cilium components
```
Comment (gandro) [inline install/kubernetes/cilium/values.yaml]: This comment implies the configuration is used for all components, but I don't think that's true? Because cilium-agent and cilium-operator run in the same network namespace (i.e. host ns), they are using a different port (6600 vs 6061). Hubble Relay also uses a different port (6062), but I think that's not strictly necessary (since it's running in its own network ns)

We should clearly document that this port is only used for the agent. Maybe it's even worth adding additional Helm flags for the operator and relay too?

Comment (chancez) [reply install/kubernetes/cilium/values.yaml]: That's what I had thought as well, but we only have a global `pprof` field in values.yaml, and from what I can tell, `cilium-operator` uses the `cilium-config` ConfigMap which is what configures it for both of them. I didn't realize that the operator ran on the host network though. I probably didn't hit any issues because I initially only tested on the agent. 

I guess I need to add a pprof section for every component so these can be configured separately.

Comment (chancez) [reply install/kubernetes/cilium/values.yaml]: Alright, I've broken out the configuration of pprof for each component. I've opted to keep the shared cilium-config configmap to ease back porting this PR.

**Comment by chancez** [comment]
As discussed in slack, marking for backport into 1.12 and 1.13 as it can help ease debugging of issues.

File: pkg/defaults/defaults.go [hunk 2]
```diff
@@ -17,6 +17,12 @@ const (
 	// ClusterMeshHealthPort is the default value for option.ClusterMeshHealthPort
 	ClusterMeshHealthPort = 80
 
+	// PprofAddressAgent is the default value for pprof in the agent
+	PprofAddressAgent = "localhost"
+
+	// PprofAddressAPIServer is the default value for pprof in the agent
```
Comment (gandro) [inline pkg/defaults/defaults.go]:
```suggestion
	// PprofAddressAPIServer is the default value for pprof in the clustermesh-apiserver
```

File: Documentation/contributing/development/dev_setup.rst [hunk 3]
```diff
@@ -656,15 +656,15 @@ These options enable logging functions in the datapath: ``cilium_dbg()``,
...
 The image below shows the options that could be used as startup options by
 ``cilium-agent`` (see upper blue box) or could be changed at runtime by running
 ``cilium config <option(s)>`` for an already running agent (see lower blue box).
 Along with each option, there is one or more logging function associated with it:
 ``cilium_dbg()`` and ``printk()``, for ``DEBUG`` and ``cilium_dbg_lb()`` for
-``DEBUG_LB``. 
+``DEBUG_LB``.
 
-.. image:: _static/cilium-debug-datapath-options.svg 
+.. image:: _static/cilium-debug-datapath-options.svg
```
Comment (gandro) [inline Documentation/contributing/development/dev_setup.rst]: While I appreciate the cleanup, those changes here are unrelated and should be in a separate commit (or even separate PR) to help with backporting conflicts and `git annotate` .

**Review by gandro — COMMENTED** [review]
Awesome, thanks for going through the effort of making the options per-component. Two small nits remaining.

**Review by gandro — APPROVED** [review]
Thanks!

**Comment by squeed** [comment]
Seems reasonable, though I will note that this doesn't work with per-node configs (since those are resolved in an InitContainer).

While this is ugly, I suspect we should just add per-process keys in the configmap. (And the world should just switch to unix domain sockets for pprof and metrics, mutter mutter grump).

**Review by jibi — APPROVED** [review]

**Comment by chancez** [comment]
> Seems reasonable, though I will note that this doesn't work with per-node configs (since those are resolved in an InitContainer).
> 
> While this is ugly, I suspect we should just add per-process keys in the configmap. (And the world should just switch to unix domain sockets for pprof and metrics, mutter mutter grump).

My thought was to separate the two configmaps for operator and agent longer term. I don't think it's going to be much fun prefixing all our values if we want to share a configmap.

**Comment by squeed** [comment]
Commenting after out-of-band discussion: If there was a way for the agent to read the pprof config directly from the configmap, it wouldn't break per-node configurations. Hopefully @chancez can find a way to make it clean :-).

**Comment by chancez** [comment]
Based on @squeed comment, I'll be switching from configuring these via env-vars to using an `operator-` prefixed option for the operator component instead, so the options can still be set in the configmap.

**Review by squeed — APPROVED** [review]
thanks!

File: clustermesh-apiserver/main.go [hunk 4]
```diff
@@ -243,6 +243,9 @@ func runApiserver() error {
 	flags.Bool(option.PProf, false, "Enable serving the pprof debugging API")
 	option.BindEnv(vp, option.PProf)
 
+	flags.String(option.PProfAddress, defaults.PprofAddressAPIServer, "Address that the pprof listens on")
```
Comment (tklauser) [inline clustermesh-apiserver/main.go]: Nit:
```suggestion
	flags.String(option.PProfAddress, defaults.PprofAddressAPIServer, "Address that pprof listens on")
```

Might as well fix it in the flag usage below for the pprof port.

File: daemon/cmd/daemon_main.go [hunk 5]
```diff
@@ -752,6 +752,9 @@ func initializeFlags() {
 	flags.Bool(option.PProf, false, "Enable serving the pprof debugging API")
 	option.BindEnv(Vp, option.PProf)
 
+	flags.String(option.PProfAddress, defaults.PprofAddressAgent, "Address that the pprof listens on")
```
Comment (tklauser) [inline daemon/cmd/daemon_main.go]: Nit:

```suggestion
	flags.String(option.PProfAddress, defaults.PprofAddressAgent, "Address that pprof listens on")
```

Same here, might change it in the pprof port message as well.

File: operator/option/config.go [hunk 6]
```diff
@@ -109,10 +109,13 @@ const (
 	OperatorPrometheusServeAddr = "operator-prometheus-serve-addr"
 
 	// PProf enabled pprof debugging endpoint
-	PProf = "pprof"
+	PProf = "operator-pprof"
```
Comment (tklauser) [inline operator/option/config.go:112]: Is there a way we could instead deprecate the old flag name for the operator and remove it in Cilium 1.14? Same for `pprof-port`/`operator-pprof-port`.

In any case we should also document the fact that the flag was rename in the version specific upgrade notes: https://docs.cilium.io/en/latest/operations/upgrade/#version-specific-notes

Comment (chancez) [reply operator/option/config.go:112]: Not really, no. The conflict means the old flag name will cause operator and agent to use the same port when reading from the configmap configfile, causing one to fail at startup. I originally avoided renaming the flag, but then the per-node config PR came up and I didn't want to have a special case here.

I can add upgrade notes though. I didn't think it would be necessary given it's updated in helm, but I guess not everyone is running via helm.

**Comment by chancez** [comment]
I've updated the flag help text + regenerated the cmdrefs. I also updated the `upgrade.rst` with notes about the changes to the cilium-operator pprof configuration flags, and mentioned the new/changed options in the helm options section.

**Comment by chancez** [comment]
Not sure why KIND is failing to start. It's failing to even bring up the control-plane, so it's not related to my PR.

File: operator/option/config.go (hunk 6 above)
Comment (tklauser) [reply operator/option/config.go:112]: Thanks for the explanation. I think this is fine given that these flags are presumably mainly used for development/debugging.

**Review by tklauser — APPROVED** [review]
Thanks for adding the upgrade notes!
